crud/common/rebalance.lua

```diff
@@ -0,0 +1,181 @@
+local fiber = require('fiber')
+local vshard_consts = require('vshard.consts')
+local utils = require('crud.common.utils')
+
+local MODULE_INTERNALS = '__module_crud_rebalance'
+local SETTINGS_SPACE_NAME = '_crud_settings'
+
+
+local M = rawget(_G, MODULE_INTERNALS)
+if not M then
+    M = {
+        safe_mode = false,
+        safe_mode_enable_hooks = {},
+        safe_mode_disable_hooks = {},
+        _router_cache_last_clear_ts = fiber.time()
+    }
+else
+    return M
+end
+
+local function create_space()
+    local settings_space = box.schema.space.create(SETTINGS_SPACE_NAME, {
+        engine = 'memtx',
+        format = {
+            { name = 'key', type = 'string' },
+            { name = 'value', type = 'any' },
+        },
+        if_not_exists = true,
+    })
+    settings_space:create_index('primary', { parts = { 'key' }, if_not_exists = true })
+end
+
+local function safe_mode_trigger(_, new, space, op)
+    if space ~= '_bucket' then
+        return
+    end
+    if (op == 'INSERT' and new.status == vshard_consts.BUCKET.RECEIVING) or
+            (op == 'REPLACE' and new.status == vshard_consts.BUCKET.SENDING) then
+        box.broadcast('_crud.safe_mode_enable', true)
+    end
+end
+
+local function register_enable_hook(func)
+    M.safe_mode_enable_hooks[func] = true
+end
+
+local function remove_enable_hook(func)
+    M.safe_mode_enable_hooks[func] = nil
+end
+
+local function register_disable_hook(func)
+    M.safe_mode_disable_hooks[func] = true
+end
+
+local function remove_disable_hook(func)
+    M.safe_mode_disable_hooks[func] = nil
+end
+
+local function safe_mode_status()
+    return M.safe_mode
+end
+
+local function safe_mode_enable()
+    if not box.info.ro then
+        box.space[SETTINGS_SPACE_NAME]:replace{ 'safe_mode', true }
+        box.space._bucket:on_replace(nil, safe_mode_trigger)
+    end
+    M.safe_mode = true
+
+    for hook, _ in pairs(M.safe_mode_enable_hooks) do
+        hook()
+    end
+end
+
+local function safe_mode_disable()
+    if not box.info.ro then
+        box.space[SETTINGS_SPACE_NAME]:replace{ 'safe_mode', false }
+        box.space._bucket:on_replace(safe_mode_trigger)
+    end
+    M.safe_mode = false
+
+    for hook, _ in pairs(M.safe_mode_disable_hooks) do
+        hook()
+    end
+end
+
+local function rebalance_init()
+    box.watch('box.status', function()
```

Comment:
box.watch was introduced https://github.com/tarantool/tarantool/releases/tag/2.10.0-beta2

need to write about it, that rebalance won't work on earlier versions